lib/gears/matcher.lua

```diff
@@ -0,0 +1,292 @@
+
+    self:add_matching_function(name, matching_fct, depends_on, precede)
+end
+
+--- Add a matching function.
+--
+-- @tparam string name The provider name. It must be unique.
+-- @tparam function callback The callback that is called to produce properties.
+-- @tparam gears.matcher callback.self The matcher object.
+-- @param callback.o The object.
+-- @tparam table callback.properties The current properties. The callback should
+--  add to and overwrite properties in this table
+-- @tparam table callback.callbacks A table of all callbacks scheduled to be
+--  executed after the main properties are applied.
+-- @tparam[opt={}] table depends_on A list of names of sources this source depends on
+--  (sources that must be executed *before* `name`.
+-- @tparam[opt={}] table precede A list of names of sources this source have a
+--  priority over.
+-- @treturn boolean Returns false if a dependency conflict was found.
+function matcher:add_matching_function(name, callback, depends_on, precede)
+    depends_on = depends_on  or {}
+    precede    = precede or {}
+    assert(type( depends_on ) == "table")
+    assert(type( precede    ) == "table")
+
+    for _, v in ipairs(self._matching_source) do
+        -- Names must be unique
+        assert(
+            v.name ~= name,
+            "Name must be unique, but '" .. name .. "' was already registered."
+        )
+    end
+
+    local new_sources = self._rule_source_sort:clone()
+
+    new_sources:prepend(name, precede    )
+    new_sources:append (name, depends_on )
+
+    local res, err = new_sources:sort()
+
+    if err then
+        gdebug.print_warning("Failed to add the rule source: "..err)
+        return false
+    end
+
+    -- Only replace the source once the additions have been proven safe
+    self._rule_source_sort = new_sources
+
+    local callbacks = {}
+
+    -- Get all callbacks for *existing* sources.
+    -- It is important to remember that names can be used in the sorting even
+    -- if the source itself doesn't (yet) exists.
+    for _, v in ipairs(self._matching_source) do
+        callbacks[v.name] = v.callback
+    end
+
+    self._matching_source = {}
+    callbacks[name] = callback
+
+    for _, v in ipairs(res) do
+        if callbacks[v] then
+            table.insert(self._matching_source, 1, {
+                callback = callbacks[v],
+                name     = v
+            })
+        end
+    end
+
+    return true
+end
+
+--- Remove a source.
+--
+-- This removes sources added with `add_matching_function` or
+-- `add_matching_rules`.
+--
+-- @tparam string name The source name.
+-- @treturn boolean If the source was removed
+function matcher:remove_matching_source(name)
+    self._rule_source_sort:remove(name)
+
+    for k, v in ipairs(self._matching_source) do
+        if v.name == name then
+            table.remove(self._matching_source, k)
+            return true
+        end
+    end
+
+    self._matching_rules[name] = nil
```

Comment:
Why set it to nil?

Reply:
If `:matches_rules()` wasn't missing a line, it would cause the removed rules to potentially still be evaluated.